### PR TITLE
Orchestrator logic updating the Goals

### PR DESCRIPTION
**Pull request by alenkacz**
Summary:
The goal of this PR is to identify the part of orchestrator that should be setting goals and implement that. The supported scenarios are described in the JIRA tickets.

JIRA issues: MARATHON-8247 MARATHON-8248

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 1]
```diff
@@ -450,7 +450,17 @@ class SchedulerActions(
...
+          def killInstances(instances: Seq[Instance]): Future[Done] = async {
+            await(if (runSpec.isResident) {
+              Future.sequence(instances.map(i => instanceTracker.goalStopped(i.instanceId)))
+            } else {
+              Future.sequence(instances.map(i => instanceTracker.goalDecommissioned(i.instanceId)))
+            })
+            await(killService.killInstances(instances, KillReason.OverCapacity))
+          }
+
+          killInstances(instances)
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: async/await

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 2]
```diff
@@ -42,6 +42,9 @@ class TaskKiller @Inject() (
           val activeInstances = foundInstances.filter(_.isActive)
 
           if (wipe) {
+            if (runSpec.isResident) {
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: see the Google Doc for wipe=true - feel free to comment there. I am still unclear here

File: src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala [hunk 3]
```diff
@@ -155,17 +155,27 @@ private class DeploymentActor(
...
+      val ScalingProposition(instancesToKill, tasksToStart) = ScalingProposition.propose(
         runningInstances, toKill, killToMeetConstraints, scaleTo, runnableSpec.killSelection)
 
-      def killTasksIfNeeded: Future[Done] = {
-        logger.debug("Kill tasks if needed")
-        tasksToKill.fold(Future.successful(Done)) { tasks =>
-          logger.debug("Kill tasks {}", tasks)
-          killService.killInstances(tasks, KillReason.DeploymentScaling).map(_ => Done)
+      def killTasksIfNeeded(): Future[Done] = {
+        def killAndChangeGoal(instances: Seq[Instance]): Future[Done] = async {
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala]: why? it's really an implementation detail of this method and I needed the def only for async/await

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala [hunk 4]
```diff
@@ -43,6 +43,15 @@ private[marathon] class InstanceUpdateOpResolver(
       case op: ReservationTimeout =>
         updateExistingInstance(op.instanceId)(updater.reservationTimeout(_, clock.now()))
 
+      case op: GoalChange =>
+        updateExistingInstance(op.instanceId)(i => {
+          val updatedInstance = i.copy(state = i.state.copy(goal = op.goal))
+          val events = InstanceChangedEventsGenerator.events(updatedInstance, task = None, clock.now(), previousCondition = Some(i.state.condition))
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: why? or what would you do instead?

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala [hunk 5]
```diff
@@ -34,7 +36,12 @@ class ScaleAppUpdateStepImpl @Inject() (
   }
 
   def calcScaleEvent(update: InstanceChange): Option[ScaleRunSpec] = {
-    if (scalingWorthy(update.condition) && update.lastState.forall(lastState => !scalingWorthy(lastState.condition))) {
+    if ((update.instance.state.goal == Goal.Running
+      // we need this to keep backward compatible with the kill&wipe=true feature for residents
+      || (update.instance.state.goal == Goal.Decommissioned && update.instance.hasReservation))
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: well it's up for discussion but after introduction of goals, I think we should trigger scale check on instance change ONLY when we expect that instance to be actually running (so goal running). Which is what I did with adding the Goal=Running condition

BUT

I think that work for all cases except wipe=true for residents, when we actually decommission but we MIGHT need to scale up.

Again this is up for discussion :)

[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: Not sure. I think on master the goals should have hardly any impact except for expunging instances when they become terminal. I'm not even sure we should considered them here. With phase 2 or 3 the scale check would go anyways.

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: Not publish any events. Are we using them anywhere?

**Comment by jeschkies** [comment]
@alenkacz, I've setup a loop [here](https://jenkins.mesosphere.com/service/jenkins/view/Marathon/job/marathon-sandbox/job/marathon-loop-PR-6328/).

File: src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala]: Testing. I just don't like nested methods so much. I guess it's because they rely on the closure capture instead of the method parameters. This makes them somewhat implicit:
```scala
def longMethod(some: String): String = {
  ...
  def bar(): String = {
     s"Where did $some come from?"
  }
  ...

  bar()
}
```

vs

```scala

def bar(input: String): String = {
  s"I'm pure and can be tested different $input"
}

def longMethod(some: String): String = {
  ...
  bar(some)
}
```

[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala]: To be fair your method has side effects and `killTasksIfNeeded()` takes no parameters.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala [hunk 6]
```diff
@@ -110,4 +110,22 @@ private[tracker] class InstanceTrackerDelegate(
 
     process(InstanceUpdateOperation.ReservationTimeout(instanceId)).map(_ => Done)
   }
+
+  override def goalDecommissioned(instanceId: Instance.Id): Future[Done] = {
```
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: Yeah, we missed to discuss it. IMHO We would have

```scala
trait Scheduler {
  def decommision(instans: Instance)
}

class InstanceTracker extends Scheduler
// or
class Scheduler(tracker: InstanceTracker)
```

[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: hmm ... what's the point of having decommision and taking instance as parameter? Does that instance already has the goal adjusted? If yes, why do we need method named `decommission`? If not, why do we pass the whole instance?

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: no... but right now it actually does not publish anything, the events generator does not contain code that would generate events for goal changes... But it can be added later. I guess I can explicitly pass empty events here...

File: src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala]: I would argue that you should not test private methods anyway, only public. And I really don't think this should be a public method.

But now I see that you are talking about `killTasksIfNeeded` while I did not introduce that method, I introduced killAndChangeGoal. But since now we are three levels deep :D I guess I an separate at least the `killTasksIfNeeded` - my `killAndChangeGoal` takes all the parameters it needs as a method should do :)

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 7]
```diff
@@ -450,7 +450,18 @@ class SchedulerActions(
             Done
         }.foreach { _ =>
           logger.info(s"Killing instances ${instances.map(_.instanceId)}")
-          killService.killInstances(instances, KillReason.OverCapacity)
+
+          def killInstances(instances: Seq[Instance]): Future[Done] = async {
+            val changeGoalsFuture = instances.map { i =>
+              if (i.hasReservation) instanceTracker.goalStopped(i.instanceId)
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: Maybe not in this PR but why is this state transition logic (*if hasReservations ... else ...) here? Imho this belongs into smth. like `Instance.transition(state)` or `Instance.update(event)` method.

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 8]
```diff
@@ -450,7 +450,18 @@ class SchedulerActions(
...
+              else instanceTracker.goalDecommissioned(i.instanceId)
+            }
+
+            await(Future.sequence(changeGoalsFuture))
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: Marathon is full of this patter - but what happens if one the instance state transitions fail e.g. because of a timeout in *InstanceTrackerDelegate*? Some of the instances will the the new `Goal` state, some wouldn't and the subsequent `killService.killInstances` will never be called.

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 9]
```diff
@@ -450,7 +450,18 @@ class SchedulerActions(
             Done
         }.foreach { _ =>
           logger.info(s"Killing instances ${instances.map(_.instanceId)}")
-          killService.killInstances(instances, KillReason.OverCapacity)
+
+          def killInstances(instances: Seq[Instance]): Future[Done] = async {
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: scale down

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 10]
```diff
@@ -42,6 +42,9 @@ class TaskKiller @Inject() (
           val activeInstances = foundInstances.filter(_.isActive)
 
           if (wipe) {
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: task kill & task kill with wipe

File: src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala [hunk 11]
```diff
@@ -155,17 +155,25 @@ private class DeploymentActor(
...
-      val ScalingProposition(tasksToKill, tasksToStart) = ScalingProposition.propose(
+      val ScalingProposition(instancesToKill, tasksToStart) = ScalingProposition.propose(
         runningInstances, toKill, killToMeetConstraints, scaleTo, runnableSpec.killSelection)
 
-      def killTasksIfNeeded: Future[Done] = {
-        logger.debug("Kill tasks if needed")
-        tasksToKill.fold(Future.successful(Done)) { tasks =>
-          logger.debug("Kill tasks {}", tasks)
-          killService.killInstances(tasks, KillReason.DeploymentScaling).map(_ => Done)
+      def killTasksIfNeeded(): Future[Done] = {
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala]: deleting runSpec

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: I think we figured it out and I updated the ticket desc

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 7 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: I don't have an idea how to do this...  but in the future we will use Stopped also for other cases, not just residents

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 8 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: hmm that's a fair point... would you retry in that case?

It actually does not make much difference now but LATER Scheduler will use the goal to actually issue the kill so then it would be super-important to actually have the goals set correctly... also then the following line with kill would be removed

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: It kind of depends how the interface should look like. As I understood @zen-dog, he likes imperatives `run`, `stop` and `decommission`. The instance tracker is rather a simple CRUD interface which we probably won't expose except for getter methods if at all. That's my take at least.

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: We'll definitely need that `InstanceStateChanged` event. Why not add it now?

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala [hunk 12]
```diff
@@ -41,7 +41,8 @@ object InstanceUpdater extends StrictLogging {
         case TaskUpdateEffect.Update(updatedTask) =>
           val updated: Instance = updatedInstance(instance, updatedTask, now)
           val events = eventsGenerator.events(updated, Some(updatedTask), now, previousCondition = Some(instance.state.condition))
-          if (updated.tasksMap.values.forall(_.isTerminal)) {
+          // TODO(alena) expunge only tasks in decommissioned state
+          if (updated.tasksMap.values.forall(_.isTerminal) && updated.state.goal != Goal.Stopped) {
```
[the remark this answers is not in the document]
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: nit: for those "complex" if-conditions I like to have one-liner `def`s with "talking names" e.g. `if (existTerminalTaskWhichAreNotStopped)` or so.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: I'd only pass the necessary information, in this case, the `instanceId` but smth. like `def decommission(instanceId)` has a "read-flow" while `goalDecommissioned(instanceId: Instance.Id)` not.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala [hunk 13]
```diff
@@ -28,9 +28,9 @@ class NotifyRateLimiterStepImpl @Inject() (
 
   override def process(update: InstanceChange): Future[Done] = {
     update.condition match {
-      case condition if limitWorthy(condition) =>
+      case condition if limitWorthy(condition) && update.stateUpdated =>
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala]: out of curiosity: what does this change do?

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 7 above)
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: If goalStopped has a side effect I'd prefer `setGoalStopped` or something

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala [hunk 14]
```diff
@@ -155,6 +155,11 @@ class TaskReplaceActor(
           logger.info(s"Killing old ${nextOldInstance.instanceId}")
       }
 
+      if (runSpec.isResident) {
+        instanceTracker.goalStopped(nextOldInstance.instanceId)
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: Why not:

```
instanceTracker.updateGoal(nextOldInstance.instanceId, Goal.Stopped)
```

?

Do we need a separate method for each goal state or sufficient to pass the appropriate enum value?

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala]: I find the condition here pretty easy to read.

File: src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala [hunk 15]
```diff
@@ -50,6 +50,12 @@ trait InstanceTracker {
   def updateStatus(instance: Instance, mesosStatus: mesos.Protos.TaskStatus, updateTime: Timestamp): Future[Done]
 
   def reservationTimeout(instanceId: Instance.Id): Future[Done]
+
+  def goalDecommissioned(instanceId: Instance.Id): Future[Done]
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/task/tracker/InstanceTracker.scala]: Seeing the `updateStatus` method above, it seems more warranted that we introduce a single `updateGoal` method with the enum. Eager to hear your thoughts on this.

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: well this is not instance state, it is Instance goal. We had a discussion about this and:
- the current InstanceChange event is very tight to the condition which does not change when goal changes
- since nobody is using these events now we decided not to introduce them at this point
- I am not that sure it makes sense to publish `InstanceGoalChanged` through API - I think that outside users should be OK with condition changes

To sum up - it's not needed right now and it's not obvious how will that event look like and who will consume it :)

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala]: yeah, well the interface looked like the one you proposed in the beginning, but then `stop` and `run` does not :( so for consistency, I had to prepend goal also to `decommission`

I am open to suggestions (that would work for all the goals)

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala (hunk 13 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala]: I am glad you asked! :)

RateLimiter is triggered for every `InstanceChange` event. Right now, `InstanceChange` is very tight to condition change so InstanceChange pretty much was condition change all the time. But now, we will be having `InstanceChange` event also for goal changes, so my idea was not to trigger rate limiter for that

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 16]
```diff
@@ -42,6 +42,9 @@ class TaskKiller @Inject() (
...
+            if (runSpec.isResident) {
+              val decommissioned = await(Future.sequence(foundInstances.map(i => instanceTracker.goalDecommissioned(i.instanceId))))
```
[the remark this answers is not in the document]
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: yes, because this is wipe=true, so for this thing, we want to wipe the resident instance so that we can unreserve in `OfferMatcherReconciler` later (basically to wipe a reservation, we need to expunge the instance because reservation is saved on instance).

With decommissioned introduced, we might actually improve this flow in the future, see https://jira.mesosphere.com/browse/MARATHON-8295

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 8 above)
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: > hmm that's a fair point... would you retry in that case?

Ideally, we **would not** retry ourselves - retry should be either user re-triggered (because the user got an exception in the API) or Mesos re-triggered (because we didn't ACK-ed the corresponding status update). But I fear that in this case, this call will end up in limbo.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Got it. Thx 👍

File: src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala (hunk 14 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala]: +1. The scheduler could have this
```scala
trait Scheduler {
  def run(instance: Instance*) : Future[Done]
  def stop(instance: Instance*) : Future[Done]
  def decommission(instance: Instance*) : Future[Done]
}
```

its implementation would then call `tracker.update` or `tracker.updateGoal`.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala (hunk 13 above)
Comment (zen-dog) [reply src/main/scala/mesosphere/marathon/core/task/update/impl/steps/NotifyRateLimiterStepImpl.scala]: so you only trigger rate limiter if the underlying condition actually change. I see. That definitely deserve an explanatory comment:
```
// RateLimiter is triggered for every InstanceChange event. Right now, InstanceChange is very tight to condition 
// change so InstanceChange pretty much was condition change all the time. But now, we will be having 
// InstanceChange event also for goal changes, so my idea was not to trigger rate limiter for that
```

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (meichstedt) [reply src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdateOpResolver.scala]: To add to this, I think this should eventually differentiate between events that inform about a goal change (aka _change of plans_) and events that inform about a conditional change (aka _something happened_). I agree that the former is not used/needed now. 

I'm not sure we should use the `InstanceChangedEventsGenerator` now if it doesn't produce events – the way it is built, it has to infer _what_ has changed, but it only has the previous _condition_. With the current signature, the generator would be completely unable to create an event that informs about a goal change, so why use it?

**Comment by jeschkies** [comment]
I rebased on master.
